st2api/st2api/controllers/v1/auth.py

```diff
@@ -0,0 +1,171 @@
+                GET /keys/
+        """
+
+        api_key_dbs = ApiKey.get_all(**kw)
+        api_keys = [ApiKeyAPI.from_model(api_key_db, mask_secrets=True)
+                    for api_key_db in api_key_dbs]
+
+        return api_keys
+
+    @request_user_has_permission(permission_type=PermissionType.API_KEY_CREATE)
+    @jsexpose(body_cls=ApiKeyAPI, status_code=http_client.CREATED)
+    def post(self, api_key_api):
+        """
+        Create a new entry or update an existing one.
+        """
+        api_key_db = None
+        try:
+            api_key_api.user = self._get_user(api_key_api)
+            api_key, api_key_hash = auth_util.generate_api_key_and_hash()
+            # store key_hash in DB
+            api_key_api.key = api_key_hash
+            api_key_db = ApiKey.add_or_update(ApiKeyAPI.to_model(api_key_api))
+        except (ValidationError, ValueError) as e:
+            LOG.exception('Validation failed for api_key data=%s.', api_key_api)
+            abort(http_client.BAD_REQUEST, str(e))
+
+        extra = {'api_key_db': api_key_db}
+        LOG.audit('ApiKey created. ApiKey.id=%s' % (api_key_db.id), extra=extra)
+
+        api_key_api = ApiKeyAPI.from_model(api_key_db)
+        # Return real api_key back to user. A one-way hash of the api_key is stored in the DB
+        # only the real value only returned at create time. Also, no masking of key here since
+        # the user needs to see this value atleast once.
+        api_key_api.key = api_key
+        return api_key_api
+
+    @request_user_has_resource_permission(permission_type=PermissionType.API_KEY_DELETE)
+    @jsexpose(arg_types=[str], status_code=http_client.NO_CONTENT)
+    def delete(self, api_key_id_or_key):
+        """
+            Delete the key value pair.
+
+            Handles requests:
+                DELETE /apikeys/1
+        """
+        api_key_db = None
+        try:
+            api_key_db = ApiKey.get_by_key_or_id(api_key_id_or_key)
+        except ApiKeyNotFoundError:
+            msg = 'ApiKey matching %s for reference and id not found.', api_key_id_or_key
+            LOG.exception(msg)
+            abort(http_client.NOT_FOUND, msg)
+
+        LOG.debug('DELETE /apikeys/ lookup with api_key_id_or_key=%s found object: %s',
+                  api_key_id_or_key, api_key_db)
+
+        try:
+            ApiKey.delete(api_key_db)
+        except Exception as e:
+            LOG.exception('Database delete encountered exception during '
+                          'delete of api_key_id_or_key="%s". ', api_key_id_or_key)
+            abort(http_client.INTERNAL_SERVER_ERROR, str(e))
+            return
+
+        extra = {'api_key_db': api_key_db}
+        LOG.audit('ApiKey deleted. ApiKey.id=%s' % (api_key_db.id), extra=extra)
+
+    def _get_user(self, api_key):
+        # If a user is provided in api_key try to use that value. In the future
+        # this behavior should change.
+        user = getattr(api_key, 'user', '')
```

Comment:
I'm a bit confused about this part - do we simply read the `user` attribute provided by the user which is included in the API model?

If so, this definitely doesn't look right - this means any user could generate a token for any other user which is definitely not a good idea :)